### PR TITLE
GH#28302: fix: authorize linked-worktree patch targets

### LINKED ISSUE #28302: OpenCode apply_patch misclassifies linked-worktree targets as canonical
## Description

OpenCode apply_patch writes are rejected as canonical even when the absolute target resolves inside a linked worktree and the pre-edit tool has just returned PASS for that exact worktree.

## Reproducer

1. Start in a canonical checkout on main.
2. Run the OpenCode `aidevops_pre_edit_check` tool with a code task; it creates a linked worktree.
3. Run the same pre-edit tool with `workdir` set to that linked worktree; observe PASS.
4. Call `apply_patch` with an absolute file path inside that linked worktree.
5. Observe: `BLOCKED by canonical write policy: canonical checkouts are read-only session mirrors`.

The session reproduced this twice with an absolute path and once with a path relative from the canonical workspace. A bounded subagent editing in the same linked worktree succeeded, proving the worktree itself was valid.

## Expected Behavior

The write guard should resolve the patch target path, its Git common directory/worktree metadata, and the validated tool `workdir`; a path inside a verified linked worktree should be writable even when the OpenCode workspace root remains the canonical checkout. Canonical paths must remain blocked.

## Files

Investigate `.agents/plugins/opencode-aidevops/quality-hooks-git-safety.mjs`, `.agents/plugins/opencode-aidevops/tools.mjs`, `.agents/plugins/opencode-aidevops/tests/test-pre-edit-tool-workdir.mjs`, and `.agents/scripts/canonical-write-policy-helper.py`. Search callers/readers before editing.

## Reference Pattern

Model target resolution on the existing explicit `workdir` handling tested by `test-pre-edit-tool-workdir.mjs`; preserve fail-closed canonical protection.

## Hazards and Compatibility

Do not weaken canonical protection, trust string path prefixes, or allow symlink escapes. Resolve real paths and Git worktree identity. Preserve behavior when a target is outside Git or the worktree no longer exists.

## Verification Before Dispatch

Add a regression test with canonical workspace root plus absolute linked-worktree patch target, a canonical negative case, and a symlink/path-traversal negative case. Run the focused OpenCode plugin tests, Python tests for the canonical helper if present, and changed-file linters.

## Acceptance Criteria

- A validated linked-worktree target is writable regardless of canonical workspace root.
- The canonical checkout remains blocked.
- Symlink and traversal attempts fail closed.
- Existing pre-edit and Git-safety tests pass.

## Workarounds Applied

The primary session used a bounded implementation subagent because direct apply_patch remained blocked after exact-worktree pre-edit validation.

---
[aidevops.sh](https://aidevops.sh) v3.32.156 plugin for [OpenCode](https://opencode.ai) v1.18.3 with gpt-5.6-sol spent 1h 57m and 331,428 tokens on this with the user in an interactive session.

### PR DESCRIPTION
**Pull request by marcusquinn**
## Summary

Resolve absolute patch targets by real Git worktree identity, allowing same-repository linked worktrees from a canonical OpenCode workspace while retaining canonical, traversal, and symlink denials.

## Files Changed

.agents/plugins/opencode-aidevops/tests/test-git-safety-gate.mjs, .agents/scripts/canonical-write-policy-helper.py, tests/test-git-safety-guard.py

## Runtime Testing

- **Risk level:** Low (agent prompts / infrastructure scripts)
- **Verification:** node --test .agents/plugins/opencode-aidevops/tests/test-git-safety-gate.mjs .agents/plugins/opencode-aidevops/tests/test-pre-edit-tool-workdir.mjs; python3 -m unittest tests.test-git-safety-guard; .agents/scripts/linters-local.sh --help

Resolves #28302

---
[aidevops.sh](https://aidevops.sh) v3.32.156 plugin for [OpenCode](https://opencode.ai) v1.18.3 with gpt-5.6-sol spent 4m and 76,935 tokens on this as a headless worker.

### CODE REVIEW

**Comment by marcusquinn** [comment]
## Completion Summary

- **What**: Resolve absolute patch targets by real Git worktree identity, allowing same-repository linked worktrees from a canonical OpenCode workspace while retaining canonical, traversal, and symlink denials.
- **Issue**: #28302
- **Files changed**: .agents/plugins/opencode-aidevops/tests/test-git-safety-gate.mjs, .agents/scripts/canonical-write-policy-helper.py, tests/test-git-safety-guard.py
- **Testing**: node --test .agents/plugins/opencode-aidevops/tests/test-git-safety-gate.mjs .agents/plugins/opencode-aidevops/tests/test-pre-edit-tool-workdir.mjs; python3 -m unittest tests.test-git-safety-guard; .agents/scripts/linters-local.sh --help
- **Key decisions**: none


---
[aidevops.sh](https://aidevops.sh) v3.32.156 plugin for [OpenCode](https://opencode.ai) v1.18.3 with gpt-5.6-sol spent 4m and 76,935 tokens on this as a headless worker.

---
Merged via PR #28305 to main.
_Merged by deterministic merge pass (pulse-wrapper.sh)._

**Review by alex-solovyev — DISMISSED** [review]
Auto-approved by pulse runner @alex-solovyev — author @marcusquinn confirmed collaborator, pre-merge gates passed.

---
[aidevops.sh](https://aidevops.sh) v3.32.145 automated scan.

File: .agents/scripts/canonical-write-policy-helper.py [hunk 1]
```diff
@@ -144,19 +144,22 @@ def check_write(cwd: str, file_path: str) -> dict[str, Any]:
...
+    elif context.classification == "canonical" and not (
+        target.classification == "linked"
+        and target.common_dir == context.common_dir
+    ):
         decision = "deny"
         reason = "canonical checkouts are read-only session mirrors"
     else:
         decision = "allow"
-        reason = "write target and process context are outside canonical worktrees"
+        reason = "write target resolves inside an allowed linked worktree"
```
[the remark this answers is not in the document]
Comment (marcusquinn) [reply .agents/scripts/canonical-write-policy-helper.py]: @gemini-code-assist 
Addressed in 135ea020d. `.agents/scripts/canonical-write-policy-helper.py:162-166` now preserves the linked-worktree reason only for linked targets and reports the outside-canonical reason otherwise. Added `tests/test-git-safety-guard.py:101-109`; `python3 -m unittest tests/test-git-safety-guard.py` passes (13 tests).

File: .agents/scripts/canonical-write-policy-helper.py [hunk 2]
```diff
@@ -185,11 +188,9 @@ def _patch_paths(patch_text: str) -> list[str]:
 
 def check_patch(cwd: str, patch_text: str) -> dict[str, Any]:
     """Return one fail-closed decision for every target in an apply patch."""
-    context_decision = check_write(cwd, "")
-    if context_decision["decision"] != "allow":
-        return context_decision
     paths = _patch_paths(patch_text)
     if not paths:
+        context_decision = check_write(cwd, "")
```
[the remark this answers is not in the document]
Comment (marcusquinn) [reply .agents/scripts/canonical-write-policy-helper.py:197]: @gemini-code-assist 
Verified this is not an unbound path. `.agents/scripts/canonical-write-policy-helper.py:195-204` returns when `paths` is empty; otherwise the non-empty list guarantees the loop at line 205 executes and binds `decision` before line 209. `python3 -m compileall -q .agents/scripts/canonical-write-policy-helper.py tests/test-git-safety-guard.py` and the focused Python/Node suites pass.

**Review by alex-solovyev — APPROVED** [review]
Auto-approved by pulse runner @alex-solovyev — author @marcusquinn confirmed collaborator, pre-merge gates passed.

---
[aidevops.sh](https://aidevops.sh) v3.32.145 automated scan.
